README.md
datalang
================

[![Travis build
status](https://travis-ci.org/edgararuiz/datalang.svg?branch=master)](https://travis-ci.org/edgararuiz/datalang)
[![Coverage
status](https://codecov.io/gh/edgararuiz/datalang/branch/master/graph/badge.svg)](https://codecov.io/github/edgararuiz/datalang?branch=master)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/datalang)](http://cran.r-project.org/package=datalang)

The goal is to help people whose first, or preferred, language is not
English. This package aids in translating the data and help files. The
hope is to lower the barrier of learning R, and also data analysis.

Even though this package can translate data on the fly, this package is
meant to be used inside another R package.

## Installation

The development version is available on [GitHub](https://github.com/):

``` r
# install.packages("devtools")
devtools::install_github("edgararuiz/datalang")
```

## How it works

The `datalang` package uses a translation spec file for translation. The
expected file format is YAML. `datalang` uses the YAML file to know
which column names, and data values, to change. The same YAML spec file
can contain the information needed to produce a help file.

Here is a sample spec file:

``` yaml
df:
  source: thisweek
  name: esta_semana
variables:
  day:
    trans: dia
    desc: Dia de la semana
    values:
      friday: viernes
      saturday: sábado
      sunday: domingo
  morning:
    trans: manana
    desc: Temperatura a las 10 AM
  afternoon:
    trans: tarde
    desc: Temperatura a las 3 PM
help:
  name: esta_semana
  alias: esta_semana
  title: Un data set de ejemplo
  description: Tres días de temperaturas tomadas en la mañana y la tarde
```

So a the data set can go from this:

    #>        day morning afternoon
    #> 1   friday      76        88
    #> 2 saturday      71        85
    #> 3   sunday      70        83

To this:

    #>       dia maniana tarde
    #> 1 viernes      76    88
    #> 2  sábado      71    85
    #> 3 domingo      70    83

## Basics

The `datalang` package includes a simple data set called: `thisweek`.
The package also has a YAML spec file with the translation.

``` r
my_spec <- system.file("specs/thisweek.yml", package = "datalang")
```

The `translate_data()` function can read the YAML spec file found in
the `my_spec` file path. It uses that spec to locate the data set, in
this case `thisweek`, modifies a copy of the data set based on the spec,
and outputs a new, translated, data set.

``` r
translate_data(my_spec)
#>       dia maniana tarde
#> 1 viernes      76    88
#> 2  sábado      71    85
#> 3 domingo      70    83
```

Not all of the spec’s entries were needed to translate the data. The
other entries are useful for creating the help documentation. In
`datalang` there are two functions that output the help:

1.  `create_rd()` - It outputs a character vector with contents that can
    be saved as an rd file.
2.  `create_html_help()` - It outputs a character vector with contents
    that can be saved as an HTML file.

## Using

There are two options to use `datalang` to provide accessibility to your
package and data:

1.  Use in a new R package dedicated to a translation to a single
    language
2.  Use to extend, and provide translations from inside an existing R
    package

### Option 1 - New package dedicated to a single translation

A package can be built to focus on translations to a single language.
Here are some of the features of this approach:

1.  The package contains copies of the data sets in the target language
2.  The created help files are R “friendly”

The `translate_folder()` function is a higher level wrapper function
that produces the translations and help files. The procedure would be to
run `translate_folder()` on the package’s project folder, and then
rebuilt the package. The default arguments have the current
expectations:

  - The translation spec YAML files will be located in the R project’s
    `inst/specs` folder.
  - The target folder to save the new data sets to will be `data`.
  - The location of the new help files will be `man`.

For a working example of this approach, check out the `datos` package.
This package is dedicated to translating data sets into Spanish:
<https://github.com/cienciadedatos/datos>.

### Option 2 - Provide translations as part of your existing package

The goal of this approach will be for the existing package to provide
the translated data, and the help files, without becoming bulky with
copies of the data. For this approach, `datalang` provides a function,
called `on_attach`, that can be run when the package is attached to the
R session.

When the package is loaded, via `library()`, then `datalang` can check
for a language setting, and then use that setting as both a trigger, and
the pointer to which language should `datalang` translate to. At that
point, the translated data set is created and loaded into the R session.

The `on_attach()` function can be added to the package’s `.onAttach()`
function, so it’s processed when the library is loaded. This is what is
currently on a fork of `gapminder`:

    .onAttach <- function(...) {
      datalang::on_attach("gapminder")
    }

Using this fork as an example. The language is set to Spanish by using
the LANGUAGE environment variable (`Sys.setenv(LANGUAGE = "es")`):

    Sys.setenv(LANGUAGE = "es")
    
    library(gapminder)
    
    #> El lenguaje asignado actualmente es español  
    #>   Cargando datos y/o ayuda de los siguientes objetos : 
    #>      gapminder  >->  paises 
    #>    Una nueva función de ayuda fue cargada: ayuda() 
    #>      Utilice asi: ayuda(paises)

#### Help function

Because the help documentation is created on-the-fly, it cannot be added
to the regular data file that functions such as `help()` use. So a new
help wrapper function is created that creates an HTML version of the
data set, or function, help. If the function, data set, or general topic
is not one created by `datalang` then the requested topic is passed
through to the `help()` function in the background.

The help’s function name will be set based on the language. So in
Spanish, the name will be: `ayuda`.


NEWS.md
# datalang 0.1

* R 4.1.0 compatibility: replace all uses of baseenv() with globalenv().

# datalang 0.0.9000

* Initial CRAN submission.


tests/testthat.R
library(testthat)
library(datalang)
library(tibble)

test_check("datalang")


tests/testthat/test-help.R
context("test-help.R")

my_spec <- system.file("specs/thisweek.yml", package = "datalang")
my_spec_path <- system.file("specs", package = "datalang")
my_translation <- system.file("translations", package = "datalang")

thisweek2 <- thisweek
thisweek2$day <- as.character(thisweek2$day)

test_that("help stuff works", {
  expect_silent(create_html_help(my_spec))
  expect_silent(datalang_help_add("thisweek", my_spec))
  expect_silent(datalang_help("thisweek"))
})


tests/testthat/test-utils.R
context("test-utils.R")

test_that("Functions in utils work", {
  expect_silent(create_this_week(path = tempdir()))
})


R/rd.R
#' Creates R help entry
#'
#' @description
#'
#' It uses the YAML spec file to create a single help entry.  Its output is a text vector.
#'
#' @param spec_path The file location of the YAML spec translation file.  It is a required argument, cannot be left empty.
#'
#' @examples
#' library(datalang)
#' my_spec <- system.file("specs/thisweek.yml", package = "datalang")
#' create_rd(my_spec)
#' @export
create_rd <- function(spec_path) {
  spec <- get_spec(spec_path = spec_path)

  header <- list(paste0("\\docType{", spec$type, "}"))
  if (!is.null(spec$help$name)) header <- c(header, paste0("\\name{", spec$help$name, "}"))
  if (!is.null(spec$help$alias)) header <- c(header, paste0("\\alias{", spec$help$alias, "}"))
  if (!is.null(spec$help$title)) header <- c(header, paste0("\\title{", spec$help$title, "}"))
  if (!is.null(spec$help$format)) header <- c(header, paste0("\\format{", spec$help$format, ""))
  header <- c(header, "\\describe{")

  items <- NULL
  items <- lapply(
    spec$variables, function(x) {
      variable <- x["trans"]
      if (variable == "TRUE") variable <- "y"
      paste0("\\item{", variable, "}{", x["desc"], "}")
    }
  )
  items <- as.character(items)
  names(items) <- NULL

  footer <- list("}}")
  if (!is.null(spec$help$usage)) {
    footer <- c(footer, paste0("\\usage{", spec$help$usage, "}"))
  }
  if (!is.null(spec$help$description)) {
    footer <- c(footer, paste0("\\description{", spec$help$description, "}"))
  }
  if (!is.null(spec$help$source)) {
    footer <- c(footer, paste0("\\source{", spec$help$source, "}"))
  }
  footer <- c(footer, "\\keyword{datasets}")


  rd <- c(header, items, footer)
  as.character(rd)
}

#' Creates and saved R help file
#'
#' @description
#'
#' Saves a the help output, from 'create_rd()', into an 'rd' file with the same name as the translated dataset.
#'
#' This function is meant for packages that will ship with a copy of the translated data set.
#'
#' @param spec_path The file location of the YAML spec translation file.  It is a required argument, cannot be left NULL.
#' @param rd_folder The target folder location where the 'rd' file will be save to. Defaults to 'man'.
#'
#' @examples
#' library(datalang)
#' my_spec <- system.file("specs/thisweek.yml", package = "datalang")
#' save_rd(my_spec, tempdir())
#' @export
save_rd <- function(spec_path = NULL, rd_folder = "man") {
  is.readable(rd_folder)

  spec <- get_spec(spec_path = spec_path)

  rd <- create_rd(spec_path)

  writeLines(
    rd,
    con = file.path(rd_folder, paste0(spec$name, ".rd")),
    useBytes = TRUE
  )
}

#' Creates rd files for multiple data sets
#'
#' @description
#'
#' Cycles through all of the spec files inside a folder specified in the 'spec_folder' argument.
#' The function translates and saves them to a specified folder.
#'
#' Saves the files to the folder specified in the 'rd_folder' argument.
#'
#' This function is meant for packages that will ship with a copy of the translated data set.
#'
#' @param spec_folder The path to the folder where the YAML spec files are located. Defaults to 'inst/specs'.
#' @param rd_folder The target folder location where the 'rd' file will be save to. Defaults to 'man'.
#'
#' @examples
#' library(datalang)
#' my_spec_folder <- system.file("specs", package = "datalang")
#' folder_rd(my_spec_folder, tempdir())
#' @export
folder_rd <- function(spec_folder = "inst/specs", rd_folder = "man") {
  specs <- get_specs_folder(spec_folder = spec_folder)

  paths <- get_item(specs, "path")

  invisible(
    lapply(paths, function(x) save_rd(x, rd_folder = rd_folder))
  )
}


R/help.R
#' Creates html help
#'
#' @description
#'
#' It uses the YAML spec file to create a single help entry.  Its output is a text vector, with 'HTML' content.
#'
#' @param spec_path The file location of the YAML spec translation file.  It is a required argument, cannot be left empty.
#' @param package Name of the package as a character variable. It is used in the help tracker.
#'
#' @examples
#' library(datalang)
#' my_spec <- system.file("specs/thisweek.yml", package = "datalang")
#' create_html_help(my_spec)
#' @export
create_html_help <- function(spec_path, package = NULL) {
  is.readable(spec_path)

  spec <- read_yaml(spec_path)

  help_name <- spec$help$name
  if (!is.null(package)) help_name <- paste0(help_name, " {", package, "}")

  h <- paste0("<body style='font-family:arial;'>")
  h <- c(h, paste0("<p>", help_name, "</p>"))
  h <- c(h, paste0("<h3>", spec$help$title, "</h3>"))
  h <- c(h, paste0("<p>", spec$help$description, "</p>"))

  h <- c(h, paste0("<table>"))
  items <- NULL
  items <- lapply(
    seq_along(spec$variables),
    function(x) {
      variable <- spec$variables[[x]]["trans"]
      description <- spec$variables[[x]]["desc"]
      if (variable == "TRUE") variable <- "y"
      if (variable == "NULL") variable <- names(spec$variables[x])
      paste0("<tr><td>", variable, "</td><td></td></tr><tr><td></td><td>", description, "</td></tr>")
    }
  )
  items <- as.character(items)
  h <- c(h, items)
  h <- c(h, paste0("</table>"))
  h <- c(h, paste0("</body>"))

  rd <- c(h)
  as.character(rd)
}

datalang_context <- new.env(parent = emptyenv())

datalang_context$help <- NULL

datalang_help_current <- function() datalang_context$help

#' Adds help entry to the current R session
#'
#' It is a mechanism to track which data sets, or functions, to display via the Help pane,
#' and which to build and display on demand.
#'
#' @param obj The object's name to be tracked
#' @param spec_path The location of the YAML spec file
#' @param package Name of the "host" package. Optional.
#'
#' @export
datalang_help_add <- function(obj, spec_path, package = NULL) {
  is.readable(spec_path)

  old <- datalang_context$help
  item <- list(
    object = obj,
    spec_path = spec_path,
    package = package
  )
  if (is.null(datalang_context$help)) {
    datalang_context$help <- list(item)
  } else {
    datalang_context$help <- c(datalang_context$help, list(item))
  }
  invisible(old)
}

#' Wrapper for 'help()' function
#'
#' @description
#'
#' Prevents unnecessary rd files to be shipped with the "host" package.  It builds and displays the data set
#' help, but it does it in the RStudio Viewer pane, as oppossed to the Help pane.  If the requested topic does
#' not match to thel help entries available via 'datalang', then 'datalang_help()' will forward the topic to
#' the regular 'help()' function.
#'
#' @param topic A quoted or unquoted name of the data set or function
#'
#' @export
datalang_help <- function(topic) {
  expr_topic <- enexpr(topic)

  dh <- datalang_help_current()
  found <- lapply(dh, function(x) x$object == expr_topic)
  found <- as.logical(found)

  if (any(found)) {
    dh_topic <- dh[found][[1]]
    ht <- create_html_help(dh_topic$spec_path, dh_topic$package)
    hs <- file.path(tempdir(), "help.html")
    writeLines(ht, hs)

    if (rstudioapi::hasFun("viewer")) {
      rstudioapi::viewer(hs)
    } else {
      browseURL(hs)
    }
  } else {
    if (inherits(expr_topic) != "character") {
      expr_topic <- as.character(expr_topic)
    }
    utils::help(topic = expr_topic)
  }
}


R/spec.R
get_spec <- function(spec_path) {
  is.readable(spec_path)

  yaml_file <- yaml::read_yaml(spec_path)

  vars <- yaml_file$variables

  variables <- lapply(
    seq_along(vars),
    function(x) {
      curr <- vars[x][[1]]
      name_curr <- names(vars[x])
      v <- list(
        name = name_curr,
        trans = ifelse(is.null(curr$trans), name_curr, curr$trans),
        desc = ifelse(is.null(curr$desc), "", curr$desc),
        has_values = !is.null(curr$values)
      )
      if (v$has_values) v$values <- curr$values
      v
    }
  )

  base <- yaml_file$df
  if (is.null(base)) base <- yaml_file

  spec <- list(
    source = base$source,
    name = base$name,
    package = ifelse(is.null(base$package), "", base$package),
    type = ifelse(is.null(base$type), "data", base$type),
    path = spec_path,
    variables = variables,
    has_help = !is.null(yaml_file$help)
  )

  if (spec$has_help) spec$help <- yaml_file$help

  spec
}

get_specs_folder <- function(spec_folder, filter_type = "") {
  is.readable(spec_folder)

  files <- list.files(spec_folder, pattern = "yml")
  files <- file.path(spec_folder, files)

  specs <- lapply(files, get_spec)

  if (filter_type != "") {
    specs_type <- lapply(specs, function(x) x$type)
    specs <- specs[specs_type == filter_type]
  }
  specs
}


R/functions.R
#' Translates and loads a function into an R environment
#'
#' @description
#'
#' Translates and loads the translation to a specified R environment.
#'
#' This function is meant for packages that will do not wish to ship a copy of the data set, and
#' wish to execute the translation on the fly, preferably when the "host" package is attached.
#'
#' @param spec_path The file location of the YAML spec translation file.  It is a required argument, cannot be left empty.
#' @param envir The target environment where the translated data set will be loaded to. Defaults to the base R environment.
#' @param package Name of the package as a character variable. It is used in the help tracker.
#'
#' @export
load_function <- function(spec_path, envir = globalenv(), package = NULL) {
  spec <- get_spec(spec_path = spec_path)

  if (spec$type != "function") stop("Not a valid function spec")

  if (is.null(package)) package <- spec$package

  fun_name <- spec$source
  if (!is.null(package)) fun_name <- paste0(package, "::", fun_name)

  a <- formals(eval(parse_expr(fun_name)))

  a_names <- get_item(spec$variables, "name")
  a_trans <- get_item(spec$variables, "trans")

  trans <- as.character(lapply(names(a), function(x) a_trans[x == a_names]))

  names(a) <- trans

  b <- lapply(
    spec$variables,
    function(x) {
      paste0(x$name, " = ", x$trans)
    }
  )
  b <- as.character(b)
  b <- paste0(b, collapse = ", ")
  b <- paste0(fun_name, "(", b, ")")
  b <- parse_expr(b)

  f <- new_function(
    eval(a),
    expr(!!b),
    envir
  )

  f <- list(f)
  names(f) <- spec$name
  env_bind(
    envir,
    !!!f
  )

  datalang_help_add(
    obj = spec$name,
    spec_path = spec_path,
    package = package
  )
}

#' Translates and loads multiple a funciton into an R environment
#'
#' @description
#'
#' Cycles through all of the spec files inside a folder specified in the 'spec_folder' argument.
#' The function translates and loads each new data set to a specified R environment.
#'
#' This function is meant for packages that will do not wish to ship a copy of the data set, and
#' wish to execute the translation on the fly, preferably when the "host" package is attached.
#'
#' @param spec_folder The path to the folder where the YAML spec files are located. Defaults to 'inst/specs'.
#' @param verbose If TRUE, it returns a list object with specs that where processed
#' @param envir The target environment where the translated data set will be loaded to. Defaults to the base R environment.
#' @param package Name of the package as a character variable. It is used in the help tracker.
#'
#' @examples
#' library(datalang)
#' my_spec_folder <- system.file("specs", package = "datalang")
#' load_folder_data(my_spec_folder)
#' @export
load_folder_functions <- function(spec_folder = "inst/specs", verbose = FALSE,
                                  envir = globalenv(), package = NULL) {
  specs <- get_specs_folder(
    spec_folder = spec_folder,
    filter_type = "function"
  )

  paths <- get_item(specs, "path")

  lapply(
    paths,
    function(x) {
      load_function(
        spec_path = x,
        envir = envir,
        package = package
      )
    }
  )

  if (verbose) specs
}


R/translate.R
#' Translates a data set
#'
#' @description
#'
#' It uses a YAML file as the spec to translate a data set.
#'
#' @param spec_path The file location of the YAML spec translation file.  It is a required argument, cannot be left empty.
#' @param .data A tibble or data.frame object that overrides the one specified in the spec file. Defaults to NULL.
#'
#' @examples
#' library(datalang)
#' my_spec <- system.file("specs/thisweek.yml", package = "datalang")
#' translate_data(my_spec)
#' @export
translate_data <- function(spec_path, .data = NULL) {
  spec <- get_spec(spec_path)

  if (spec$type != "data") stop("Not a valid data set spec")

  if (is.null(.data)) {
    df <- parse_expr(spec$source)
    df <- eval(df)
  } else {
    df <- .data
  }

  was_tibble <- is_tibble(df)
  df <- as_tibble(df)

  vars <- spec$variables

  dfl <- lapply(
    seq_along(vars),
    function(x) {
      curr <- vars[[x]]
      cl <- df[, curr$name][[1]]
      from <- names(curr$values)
      to <- as.character(vars[[x]]$values[from])
      if (!is.null(from)) {
        if ("factor" %in% class(cl)) {
          lv <- levels(cl)
          for (i in seq_along(from)) {
            lv[lv == from[i]] <- to[i]
          }
          levels(cl) <- lv
        } else {
          for (i in seq_along(from)) {
            cl[cl == from[i]] <- to[i]
          }
        }
      }

      cl <- as.data.frame(cl)
      colnames(cl) <- curr$trans
      cl
    }
  )

  dfl <- as.data.frame(dfl)

  colnames(dfl) <- sapply(
    seq_along(vars),
    function(x) { vars[[x]]$trans }
  )

  if (was_tibble) dfl <- as_tibble(dfl)

  dfl
}

#' Translates and saves a data set
#'
#' @description
#'
#' Saves a translated version of the data set in 'rda' format.  It wraps the translate_data()
#' function, and saves the output.
#'
#' This function is meant for packages that will ship with a copy of the translated data set.
#'
#' @param spec_path The file location of the YAML spec translation file.  It is a required argument, cannot be left NULL.
#' @param data_folder The target folder location where the 'rda' file will be save to.
#'
#' @examples
#' library(datalang)
#' my_spec <- system.file("specs/thisweek.yml", package = "datalang")
#' save_translation(my_spec, tempdir())
#' @export
save_translation <- function(spec_path, data_folder = "data") {
  spec <- get_spec(spec_path = spec_path)

  df_name <- spec$name

  df <- translate_data(spec_path)

  assign(df_name, df)
  save(
    list = df_name,
    file = paste0(data_folder, "/", df_name, ".rda"),
    compress = "xz"
  )
}

#' Translates and loads a data set into an R environment
#'
#' @description
#'
#' Translates and loads the translation to a specified R environment.
#'
#' This function is meant for packages that will do not wish to ship a copy of the data set, and
#' wish to execute the translation on the fly, preferably when the "host" package is attached.
#'
#' @param spec_path The file location of the YAML spec translation file.  It is a required argument, cannot be left empty.
#' @param envir The target environment where the translated data set will be loaded to. Defaults to the base R environment.
#' @param package Name of the package as a character variable. It is used in the help tracker.
#'
#' @examples
#' library(datalang)
#' my_spec <- system.file("specs/thisweek.yml", package = "datalang")
#' load_translation(my_spec)
#' @export
load_translation <- function(spec_path, envir = globalenv(), package = NULL) {
  spec <- get_spec(spec_path = spec_path)

  df <- translate_data(spec_path = spec_path)

  assign(
    x = spec$name,
    value = df,
    envir = envir
  )

  datalang_help_add(
    obj = spec$name,
    spec_path = spec_path,
    package = package
  )
}

#' Translates and loads multiple a data sets into an R environment
#'
#' @description
#'
#' Cycles through all of the spec files inside a folder specified in the 'spec_folder' argument.
#' The function translates and loads each new data set to a specified R environment.
#'
#' This function is meant for packages that will do not wish to ship a copy of the data set, and
#' wish to execute the translation on the fly, preferably when the "host" package is attached.
#'
#' @param spec_folder The path to the folder where the YAML spec files are located. Defaults to 'inst/specs'.
#' @param verbose If TRUE, it returns a list object with specs that where processed
#' @param envir The target environment where the translated data set will be loaded to. Defaults to the base R environment.
#' @param package Name of the package as a character variable. It is used in the help tracker.
#'
#' @examples
#' library(datalang)
#' my_spec_folder <- system.file("specs", package = "datalang")
#' load_folder_data(my_spec_folder)
#' @export
load_folder_data <- function(spec_folder = "inst/specs", verbose = FALSE,
                             envir = globalenv(), package = NULL) {
  specs <- get_specs_folder(
    spec_folder = spec_folder,
    filter_type = "data"
  )

  paths <- as.character(
    lapply(
      specs,
      function(x) x$path
    )
  )

  invisible(
    lapply(
      paths,
      function(x) {
        load_translation(
          spec_path = x,
          envir = envir,
          package = package
        )
      }
    )
  )

  if (verbose) specs
}

#' Translates and loads multiple a data sets into an R environment
#'
#' @description
#'
#' Cycles through all of the spec files inside a folder.  The default behaivor is if a language
#' setting is detected ('Sys.getenv("LANGUAGE")'), it will look for a sub-folder that matches the
#' setting inside the 'spec_folder' path.
#'
#' The function translates and loads each new data set to a specified R environment.
#'
#' This function is meant for packages that will do not wish to ship a copy of the data set, and
#' wish to execute the translation on the fly, preferably when the "host" package is attached.
#'
#' @param spec_folder The path to the folder where the YAML spec files are located. Defaults to 'inst/specs'.
#' @param verbose Prints to the console the name of the original data set, and the name of the new translated data set.
#' @param envir The target environment where the translated data set will be loaded to. Defaults to the base R environment.
#' @param language Optional argument that allows the target language to be specified.
#' @param package Name of the package as a character variable. It is used in the help tracker.
#'
#' @export
load_package_translations <- function(spec_folder = "translations",
                                      verbose = TRUE,
                                      envir = globalenv(),
                                      language = NULL,
                                      package = NULL) {
  if (is.null(language)) language <- Sys.getenv("LANGUAGE")

  if (language != "") {
    lang_folder <- file.path(spec_folder, language)
    if (file.exists(lang_folder)) {
      specs <- get_specs_folder(
        spec_folder = lang_folder
      )

      max_source <- max(as.integer(lapply(specs, function(x) nchar(x$source))))
      max_source <- max_source + 2

      types <- as.character(lapply(specs, function(x) x$type))

      msgs <- get_messages(language)
      startup <- NULL
      startup <- c(startup, paste0(msgs$startup$detected))
      startup <- c(startup, paste0(msgs$startup$datasets, ":"))

      if (any(types == "data")) {
        load_folder_data(
          lang_folder,
          verbose = verbose,
          envir = envir,
          package = package
        )
        items <- specs[types == "data"]
        m <- as.character(
          lapply(
            items,
            function(x) {
              ifelse(x$source == x$name, x$name, paste0(pad_str(x$source, max_source, "-"), "> ", x$name))
            }
          )
        )
        startup <- c(startup, paste0(" ", msgs$startup$data))
        startup <- c(startup, paste0("  ", m))
      }

      if (any(types == "function")) {
        load_folder_functions(
          lang_folder,
          envir = envir,
          verbose = verbose
        )
        items <- specs[types == "function"]
        m <- as.character(
          lapply(
            items,
            function(x) {
              ifelse(x$source == x$name, x$name, paste0(pad_str(x$source, max_source, "-"), "> ", x$name))
            }
          )
        )
        startup <- c(startup, paste0(" ", msgs$startup$functions))
        startup <- c(startup, paste0("  ", m))
      }

      create_help_function(name = msgs$help$name)
      startup <- c(startup, paste0(msgs$help$message, ": ", msgs$help$name, "()"))
      startup <- c(startup, paste0(msgs$help$use, ": ", msgs$help$name, "(", specs[[1]]$name, ")"))


      startup <- paste0(startup, collapse = "\n")
      if (verbose) cat(startup)
    }
  }
}

#' @inherit load_package_translations
#' @export
on_attach <- function(package = NULL,
                      spec_folder = system.file("translations", package = package),
                      envir = as.environment(paste0("package:", package)),
                      language = NULL,
                      verbose = TRUE) {
  load_package_translations(
    spec_folder = spec_folder,
    envir = envir,
    package = package,
    language = language
  )
}

#' Translates and saves multiple a data sets
#'
#' @description
#'
#' Cycles through all of the spec files inside a folder specified in the 'spec_folder' argument.
#' The function translates and saves them to a specified folder.
#'
#' Saves a translated versions of the data set in 'rda' format.  It wraps the translate_data()
#' function, and saves the output.
#'
#' This function is meant for packages that will ship with a copy of the translated data set.
#'
#' @param spec_folder The path to the folder where the YAML spec files are located. Defaults to 'inst/specs'.
#' @param data_folder The target folder location where the 'rda' file will be save to.
#'
#' @examples
#' library(datalang)
#' my_spec_folder <- system.file("specs", package = "datalang")
#' folder_data(my_spec_folder, tempdir())
#' @export
folder_data <- function(spec_folder = "inst/specs", data_folder = "data") {
  specs <- get_specs_folder(
    spec_folder = spec_folder,
    filter_type = "data"
  )

  paths <- as.character(
    lapply(
      specs,
      function(x) x$path
    )
  )

  lapply(
    paths,
    function(x) {
      save_translation(spec_path = x, data_folder = data_folder)
    }
  )
}

#' Translates and saves multiple a data sets
#'
#' @description
#'
#' Cycles through all of the spec files inside a folder specified in the 'spec_folder' argument.
#' The function translates and saves them to a specified folder. It also creates and saves the
#' help files.
#'
#' Saves a translated versions of the data set in 'rda' format.
#'
#' This function is meant for packages that will ship with a copy of the translated data set.
#'
#' @param spec_folder The path to the folder where the YAML spec files are located. Defaults to 'inst/specs'.
#' @param data_folder The target folder location where the 'rda' file will be saved to.
#' @param rd_folder The target folder location where the help files will be saved to.
#'
#' @examples
#' library(datalang)
#' my_spec_folder <- system.file("specs", package = "datalang")
#' folder_data(my_spec_folder, tempdir())
#' @export
translate_folder <- function(spec_folder = "inst/specs",
                             data_folder = "data", rd_folder = "man") {
  folder_data(spec_folder, data_folder)
  folder_rd(spec_folder, rd_folder)
}


R/utils.R
create_this_week <- function(path = "data") {
  thisweek <- data.frame(
    day = c("friday", "saturday", "sunday"),
    morning = c(76, 71, 70),
    afternoon = c(88, 85, 83)
  )
  save(thisweek, file = file.path(path, "thisweek.rda"))
}

get_messages <- function(language = "es") {
  path <- system.file("messages", package = "datalang")
  msgs <- list.files(path)
  msgs <- msgs[msgs == paste0(language, ".yml")]
  spec <- file.path(path, msgs)
  if (length(msgs) == 0) spec <- file.path(path, "en.yml")
  yaml::read_yaml(spec)
}

create_help_function <- function(name,
                                 message = "New help function added:",
                                 usage = "Usage:",
                                 example = "mtcars") {
  f <- new_function(
    alist(... = ),
    quote(datalang::datalang_help(...))
  )
  f <- list(f)
  names(f) <- name
  env_bind(base_env(), !!!f)
  message <- paste0(message, ":")
  cat("  ", message, paste0(name, "()"), "\n")
  cat("    ", paste0(usage, ": ", name, "(", example, ")"), "\n")
}
